Common/GPU/Vulkan/VulkanQueueRunner.cpp

```diff
@@ -1526,21 +1535,68 @@ void VulkanQueueRunner::PerformRenderPass(const VKRStep &step, VkCommandBuffer c
 
 		case VKRRenderCommand::DRAW_INDEXED:
 		{
+			static const char indexShift[2] = { 1, 2 };
+
 			vkCmdBindDescriptorSets(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, pipelineLayout, 0, 1, &c.drawIndexed.ds, c.drawIndexed.numUboOffsets, c.drawIndexed.uboOffsets);
-			vkCmdBindIndexBuffer(cmd, c.drawIndexed.ibuffer, c.drawIndexed.ioffset, (VkIndexType)c.drawIndexed.indexType);
-			VkDeviceSize voffset = c.drawIndexed.voffset;
-			vkCmdBindVertexBuffers(cmd, 0, 1, &c.drawIndexed.vbuffer, &voffset);
-			vkCmdDrawIndexed(cmd, c.drawIndexed.count, c.drawIndexed.instances, 0, 0, 0);
+			int indexOffset = 0;
+
+			// Conservative limits for bind eliminination for Adreno. Still a good effect.
+			if (lastIndexOffset + c.drawIndexed.count > 32768) {
+				lastIndexBuffer = 0;
+			}
+
+			if (c.drawIndexed.ibuffer == lastIndexBuffer && c.drawIndexed.indexType == lastIndexType) {
+				indexOffset = (c.drawIndexed.ioffset - lastIndexOffset) >> (c.drawIndexed.indexType + 1);  // NOTE: This only works for UINT16 = 0 and UINT32 = 1.
+			} else {
+				vkCmdBindIndexBuffer(cmd, c.drawIndexed.ibuffer, c.drawIndexed.ioffset, (VkIndexType)c.drawIndexed.indexType);
+				lastIndexBuffer = c.drawIndexed.ibuffer;
+				lastIndexType = (VkIndexType)c.drawIndexed.indexType;
+				lastIndexOffset = c.drawIndexed.ioffset;
+			}
+
+			// Conservative limits for bind elimination for Adreno. Still a good effect.
+			int vertexOffset = 0;
+			if (lastVertexOffset > 16834) {
+				lastVertexBuffer = 0;
+			}
+
+			if (c.drawIndexed.vbuffer == lastVertexBuffer && (c.drawIndexed.voffset - lastVertexOffset) % lastGraphicsPipeline->vertexStride == 0) {  // This is invalidated on pipeline binds, so vertex stride will match
+				vertexOffset = (c.drawIndexed.voffset - lastVertexOffset) / lastGraphicsPipeline->vertexStride;
+			} else {
+				VkDeviceSize voffset = c.drawIndexed.voffset;
+				vkCmdBindVertexBuffers(cmd, 0, 1, &c.drawIndexed.vbuffer, &voffset);
+				lastVertexBuffer = c.drawIndexed.vbuffer;
+				lastVertexOffset = voffset;
+			}
+
+			vkCmdDrawIndexed(cmd, c.drawIndexed.count, c.drawIndexed.instances, indexOffset, vertexOffset, 0);
 			break;
 		}
 
 		case VKRRenderCommand::DRAW:
+		{
 			vkCmdBindDescriptorSets(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, pipelineLayout, 0, 1, &c.draw.ds, c.draw.numUboOffsets, c.draw.uboOffsets);
+
+			// Conservative limits for bind elimination for Adreno. Still a good effect.
+			int vertexOffset = 0;
+			if (lastVertexOffset > 16834) {
```

Comment:
Won't this be most of the time?  I thought we rolled forward in push buffers, such that it starts at 0 only at the beginning of the frame.

Is it an issue for `c.draw.voffset - lastVertexOffset > 16384`?

Also, we may want to check for wrap-around.  Maybe VMA is giving us the same underlying buffer, but different checks.  I bet `vertexOffset` isn't allowed to be negative?

-[Unknown]

Reply:
We do, and for some reason I thought I was checking the relative step from the last bind here, but in reality I'm checking the absolute offset, so this is pretty pointless indeed right now.

Could fix that and see if it worked, but I think I'm just gonna give up on this change on Adreno, and possibly entirely.